keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+from keras_nlp.layers.modeling.rotary_embedding import RotaryEmbedding
+from keras_nlp.models.llama.dynamic_ntk_scaled_rotary_embedding import (
+    DynamicNTKScaledRotaryEmbedding,
+)
+from keras_nlp.utils.keras_utils import clone_initializer
+
+
+class LlamaAttention(keras.layers.Layer):
+    def __init__(
+        self,
+        num_heads,
+        hidden_dim,
+        num_key_value_heads,
+        rope_scaling_type=None,
+        rope_scaling_factor=None,
+        dropout=0.0,
+        rotary_percentage=0.25,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        rotary_max_wavelength=10000,
+        max_sequence_length=512,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.num_heads = num_heads
+        self.num_key_value_heads = num_key_value_heads
+
+        self.num_key_value_groups = num_heads // num_key_value_heads
+        self.attn_head_size = hidden_dim // num_heads
+        self.rotary_max_wavelength = rotary_max_wavelength
+
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+        self.max_sequence_length = max_sequence_length
+
+        self.rotary_dim = int(self.attn_head_size * rotary_percentage)
+        self.rope_scaling_type = rope_scaling_type
+        self.rope_scaling_factor = rope_scaling_factor
+
+        self.dropout = dropout
+
+        if rope_scaling_type is None:
```

Comment:
Keep in mind it is usually easiest to just support a minimal set of options we need for the checkpoints of interest. Are all three of these options used across different checkpoints?

Reply:
They are not being used in public checkpoints.

Reply:
I don't think that it's much of a deal which makes the implementation complex!

Reply:
Whatever options aren't used in public checkpoints let's just ditch. We won't really be able to verity it's working right without checkpoints anyway.

Which are the options that are needed currently?

Reply:
I have removed DynamicScaledNTKRotary Embedding to as none of the publically provided checkpoints by llama2 use it.